ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Protocol/Praos.hs

```diff
@@ -118,52 +120,65 @@ praosValidateView getFields hdr =
   Forging
 -------------------------------------------------------------------------------}
 
-newtype HotKey c = HotKey {
-      getHotKey :: SignKeyKES (PraosKES c)
-    }
+data HotKey c =
+    HotKey
+      !Period  -- ^ Absolute period of the KES key
+      !(SignKeyKES (PraosKES c))
+  | HotKeyPoisoned
   deriving (Generic)
 
-deriving instance PraosCrypto c => Show (HotKey c)
+instance PraosCrypto c => NoUnexpectedThunks (HotKey c)
 
--- We override 'showTypeOf' to make sure to show @c@
-instance PraosCrypto c => NoUnexpectedThunks (HotKey c) where
-  showTypeOf _ = show $ typeRep (Proxy @(HotKey c))
+deriving instance PraosCrypto c => Show (HotKey c)
 
-evolveKey :: PraosCrypto c => SlotNo -> HotKey c -> HotKey c
-evolveKey slotNo (HotKey oldKey) =
-    HotKey $ fromMaybe (error "evolveKey: updateKES failed") $
-               updateKES () oldKey kesPeriod
+-- | The 'HotKey' could not be evolved to the given 'Period'.
+newtype HotKeyEvolutionError = HotKeyEvolutionError Period
+  deriving (Show)
+
+-- | To be used in conjunction with, e.g., 'updateMVar'.
+evolveKey ::
```

Comment:
Comment: when the key's period is after the target period, we shouldn't use it, but we do. That's what we check in `tpraosCheckCanForge` in the real implementation.